Adding more structure to Shoot Cluster Maintenance (#111)
gardener/documentation: website/documentation/050-tutorials/content/howto/maintain-shoot/_index.md

# Shoot Cluster Maintenance

Day two operations for shoot clusters are related to:

> * The Kubernetes version of the control planes and the worker nodes and

Reviewer: ```suggestion
* the Kubernetes version of the control plane and the worker nodes
```

If you don’t want to wait for the next maintenance window, you can annotate the `Shoot` specification with `shoot.gardener.cloud/operation: maintain`. Gardener then checks immediately if there’s an auto-update or a forceful update needed.

{{% notice info %}}
<p>
Forceful version updates are even executed if the auto-update for the Kubernetes version, or the auto-update for the machine image version is deactivated (set to `false`).
</p>
{{% /notice %}}

With expiration dates, administrators can give shoot cluster owners more time for testing before the actual version update happens, which allows smoother transitions to new versions.

### Kubernetes Update Paths

> The bigger the delta of the Kubernetes source version and the Kubernetes target version, the better it must be planned and executed by operators. Gardener only provides automatic support for updates that are uncritical for your deployed resources:

Reviewer: critical != harmful, patch update can be critical (for example includes CVE fixes).

```suggestion
The bigger the delta of the Kubernetes source version and the Kubernetes target version, the better it must be planned and executed by operators. Gardener only provides automatic support for updates that are not harmful for the cluster workload:
```

Author: @ialidzhikov: Thanks for your review (working on it)!

"harmful" does sound weird (I checked with a native speaker), but I see your point.

Proposal:

Gardener only provides automatic support for updates that can be applied safely to the cluster workload.

Other idea: ...that can be applied without disrupting the cluster workload...

> The bigger the delta of the Kubernetes source version and the Kubernetes target version, the better it must be planned and executed by operators. Gardener only provides automatic support for updates that can be applied safely to the cluster workload:

Author: Is "can be applied safely" o.k.?

| Other | `1.10.12` to `1.12.0`, or `1.10.12` to `2.0.0` | manual update |

Gardener doesn’t support automatic updates of major Kubernetes versions or nonconsecutive minor versions, because Kubernetes doesn’t guarantee updateability in this case. However, multiple minor version updates are possible if not only the minor source version is expired, but also the minor target version is expired. Gardener then updates the Kubernetes version first to the expired target version, and waits for the next maintenance window to update this version to the next minor target version.

{{% notice warning %}}
<p>
The administrator who maintains the `CloudProfile` has to ensure that the list of Kubernetes versions consists of consecutive minor versions, for example, from `1.10.x` to `1.11.y`. If the minor version increases in bigger steps, for example, from `1.10.x` to `1.12.y`, shoot clusters updates fail during the maintenance window. 
</p>
{{% /notice %}}

## Manual Updates

To update the Kubernetes version or the node operating system manually, change the `.spec.kubernetes.version` field or the `.spec.provider.workers.machine.image.version` field correspondingly.  

> Manual updates are required if you would like to do a minor update or a major update of the Kubernetes version. Gardener doesn’t do such updates automatically as they can have breaking changes that could impact your deployed resources.

Reviewer: ```suggestion
Manual updates are required if you would like to do a minor update of the Kubernetes version. Gardener doesn’t do such updates automatically as they can have breaking changes that could impact the cluster workload.
```

Manual updates are either executed immediately (default) or can be confined to the maintenance time window.  
Choosing the latter option, causes changes to the cluster (for example, node pool rolling-updates) and the subsequent reconciliation, to only predictably happen during a defined time window (available since [Gardener version 1.4](https://github.com/gardener/gardener/releases/tag/v1.4.0)).

More information: [Confine Specification Changes/Update Roll Out](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_maintenance.md#confine-specification-changesupdates-roll-out).

{{% notice warning %}}
<p>
Before applying such update on minor or major releases, operators should check for all the breaking changes introduced in the target Kubernetes release changelog.
</p>
{{% /notice %}}

## Examples

In the examples for the `CloudProfile` and the `Shoot` specification, only the fields relevant for the example are shown.

### Auto-Update of Kubernetes Version

Let's assume Kubernetes version `1.10.5` and `1.11.0` were added in the following `CloudProfile`:

```yaml
spec:
  kubernetes:
    versions:
    - version: 1.11.0
    - version: 1.10.5
    - version: 1.10.0
```

> Before this change, your shoot cluster specification looked like this:

Reviewer: ```suggestion
Before this change, the shoot cluster specification looked like this:
```
